docs/design/proposals/resource-reference/resource-reference.md

```diff
@@ -0,0 +1,551 @@
+* Add support for not deleting a resource if it is referenced in another resource
+* Add support for finite retries and terminal conditions when resolving references
+
+## Preferred Approach
+
+### High Level Overview
+
+* Introduce new construct in generator.yaml which will indicate to code-generator
+if a new spec field with type `AWSResourceReference` should be added to the custom
+resource definition.
+
+* This new spec field will work as additional input for an existing spec field.
+Addition of the extra spec field keeps the backward compatibility with existing
+custom resource definitions.
+
+* Validation will be present that only one of existing spec field OR new reference
+field is present in custom resource, not both.
+
+* If the existing spec field was marked as required, validation will be added that
+at least one of existing spec field or new reference field is present.
+
+* If `AWSResourceReference` type field is present in the desired state, first ACK
+controller will query for the referenced resource. If the referenced resource
+has `ACK.ResourceSynced` condition status `True` & specified field is found
+in the referenced resource, the reconciliation loop will progress. Otherwise
+the request will get requeued for future processing.
+
+* Watchers for the referenced resource type will not be implemented in the first
+revision. The first version of this implementation will mostly be used for finding
+identifier of referenced resource and any update in referenced resource should not
+change it's identifier. As mentioned earlier, deletion will also be handled in
+coming revisions.
+
+* The code to determine when the referenced resource is synced will be generated,
+and additional hooks will be added for customization. Ex: Do not just check presence
+of a field but also check if the Status is available etc...
+
+### Low Level Implementation
+
+#### generator.yaml Changes
+
+```go
+type FieldConfig struct {
+    ...
+    Reference *ReferenceConfig `json:"references,omitempty"`
+    ...
+}
+
+// ReferenceConfig contains the instructions for how to add the referenced resource
+// configuration for a field.
+type ReferenceConfig struct {
+	// Kind mentions the K8s Kind of referenced resource
+	Kind string `json:"kind"`
+	// FieldPath refers to the the path of field which should be copied
+	// from referenced resource into TargetFieldName
+	FieldPath string `json:"field_path"`
+	// TargetFieldName is spec field name which gets value from
+	// referenced resource's FieldPath
+	TargetFieldName string `json:"target_field"`
+	// IsList mentions whether the new field should be AWSResourceReference or
+	// List of AWSResourceReference
+	IsList          bool   `json:"is_list"`
+	// Required mentions whether either of AWSResourceReference or
+	// TargetFieldName must be present in custom resource
+	Required        bool   `json:"required"`
+}
+```
+
+`APIGatewayv2 generator.yaml` Example:
+```yaml
+resources:
+  Integration:
+    fields:
```

Comment:
This may be outside the scope of this proposal, but any thoughts in making `fields` more explicit such as `spec_fields` ?


The mixing of these different fields is getting confusing. This override requires knowledge of CRDs Spec/Status field paths already, so I think it'd be clearer to do something like:

```
resources:
  Integration:
    spec_fields:
      APIID:
        references:
          kind: API
          value_from_path: Status.APIID
          is_list: false
          required: true
```